### PR TITLE
Inline the usage of `nix::readDirectory` and remove it

### PR DESCRIPTION
**Pull request by siddhantk232**
`nix::readDirectory` is removed. `std::filesystem::directory_iterator` is used directly in places that used this util.

# Motivation

The `nix::readDirectory` uses `std::filesystem::directory_iterator` and pushes all the entries in a `std::vector` before returning it. This is redundant as we can directly use `std::filesystem::directory_iterator` to iterate through entries and count files (and dirs) in path.

# Context

https://github.com/NixOS/nix/issues/9205

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libstore/unix/builtins/unpack-channel.cc [hunk 1]
```diff
@@ -21,10 +21,12 @@ void builtinUnpackChannel(
 
     unpackTarfile(src, out);
 
-    auto entries = readDirectory(out);
-    if (entries.size() != 1)
+    auto entries = std::filesystem::directory_iterator{out};
+    auto file_count = std::distance(entries, std::filesystem::directory_iterator{});
```
Comment (edolstra) [inline src/libstore/unix/builtins/unpack-channel.cc:25]: I'm not sure if this is well-defined behaviour. `std::distance` will advance a copy of the `entries` iterator, reading the directory and (presumably) discarding the entries read. So I'm not sure if you can still read from `entries` after that.

There is a failure in `tests/functional/nix-channel.sh` that might be related to this.

**Review by edolstra — APPROVED** [review]
Looks good apart from the test failure.

File: src/libstore/unix/builtins/unpack-channel.cc (hunk 1 above)
Comment (siddhantk232) [reply src/libstore/unix/builtins/unpack-channel.cc]: Thanks for the help. This is fixed in https://github.com/NixOS/nix/pull/10684/commits/4d0777ca69463cbaba3501c10d16012101bb0933 and tests are now pasing!

**Comment by edolstra** [comment]
Can you resolve the conflicts? Then we can merge this.

**Comment by Ericson2314** [comment]
@siddhantk232 can you make sure ever for loop callsite has a "check interrupt" like before? This will make sure we don't regress with control-C support and very wide directories.

**Comment by siddhantk232** [comment]
> @siddhantk232 can you make sure ever for loop callsite has a "check interrupt" like before? This will make sure we don't regress with control-C support and very wide directories.

Thanks for pointing this out! I'll open another PR that does this.

File: src/nix/prefetch.cc [hunk 2]
```diff
@@ -115,9 +115,10 @@ std::tuple<StorePath, Hash> prefetchFile(
 
             /* If the archive unpacks to a single file/directory, then use
                that as the top-level. */
-            auto entries = readDirectory(unpacked);
-            if (entries.size() == 1)
-                tmpFile = entries[0].path();
+            auto entries = std::filesystem::directory_iterator{unpacked};
+            auto file_count = std::distance(entries, std::filesystem::directory_iterator{});
```
Comment (Mic92) [inline src/nix/prefetch.cc:119]: This broke the prefetch command if unpack is used. Fixed in https://github.com/NixOS/nix/pull/11051

File: src/libstore/unix/builtins/unpack-channel.cc [hunk 3]
```diff
@@ -21,10 +21,13 @@ void builtinUnpackChannel(
 
     unpackTarfile(src, out);
 
-    auto entries = readDirectory(out);
-    if (entries.size() != 1)
+    auto entries = std::filesystem::directory_iterator{out};
+    auto fileName = entries->path().string();
```
Comment (Mic92) [inline src/libstore/unix/builtins/unpack-channel.cc:25]: Maybe this way around it actually works.

**Pull request merged**